File: Placement Sheet/LinkedList/BS/NthNodeFromLast.ts
class NodeLL {
    private data: number;
    private next: NodeLL | null = null;
    constructor(data?: number) {
        if (data) {
            this.data = data;
        }
    }

    getData(): number {
        return this.data;
    }

    getNext(): NodeLL | null {
        return this.next;
    }

    setNext(node: NodeLL): void {
        this.next = node;
    }
}

class NthNodeFromLast {
    private head: NodeLL | null = null;
    private tail: NodeLL | null = null;

    insert(data: number): void {
        let newNode: NodeLL | null = new NodeLL(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            return;
        }
        this.tail?.setNext(newNode);
        this.tail = newNode;
    }

    display(): void {
        if (this.head == null) {
            console.log("Lists is empty...");
            return;
        }
        let currNode: NodeLL | null = this.head;
        let result: string = "";
        while (currNode !== null) {
            result += currNode.getData() + " --> ";
            currNode = currNode.getNext();
        }
        result += "NULL";
        console.log(result);
    }

    nthNodeFromLast(n: number): number | null {
        if (this.head == null) {
            console.log("List is empty...");
            return null;
        }
        let fastPtr: NodeLL | null = this.head;
        let slowPtr: NodeLL | null = this.head;
        for (let i: number = 0; i < n; i++) {
            if (fastPtr == null) {
                return null;
            }
            fastPtr = fastPtr.getNext();
        }
        while (fastPtr !== null) {
            fastPtr = fastPtr.getNext();
            slowPtr = slowPtr?.getNext()!;
        }
        return (slowPtr !== null) ? slowPtr.getData() : null;
    }
}

let list: NthNodeFromLast = new NthNodeFromLast();
list.insert(1);
list.insert(2);
list.insert(3);
list.insert(4);
list.insert(5);
list.insert(6);
list.insert(7);
list.display();
console.log(`Last n'th node: ${list.nthNodeFromLast(2)}`);

export { };

File: Placement Sheet/LinkedList/BS/NthNodeFromLast.java
public class NthNodeFromLast {
    private Node head;
    private Node tail;

    public void insert(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            return;
        }
        newNode.next = tail.next;
        tail.next = newNode;
        tail = newNode;
    }

    public void display() {
        if (head == null) {
            System.out.println("List is empty...");
            return;
        }
        Node currNode = head;
        while (currNode != null) {
            System.out.print(currNode.data + " --> ");
            currNode = currNode.next;
        }
        System.out.println("NULL");
    }

    public int nthNodeFromLast(int n) {
        if (head == null) {
            System.out.println("List is empty...");
            return -1;
        }
        Node fastPtr = head;
        Node slowPtr = head;
        for (int i = 0; i < n; i++) {
            if (fastPtr == null) {
                return -1;
            }
            fastPtr = fastPtr.next;
        }
        while (fastPtr != null) {
            fastPtr = fastPtr.next;
            slowPtr = slowPtr.next;
        }
        return (slowPtr != null) ? slowPtr.data : -1;
    }

    public static void main(String[] args) {
        NthNodeFromLast list = new NthNodeFromLast();
        list.insert(1);
        list.insert(2);
        list.insert(3);
        list.insert(4);
        list.insert(5);
        list.insert(6);
        list.insert(7);
        list.display();
        System.out.println("n'th node from last: " + list.nthNodeFromLast(2));
    }
}